bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,244 @@
+ * questions.
+ */
+package org.openjdk.skara.bots.topological;
+
+import org.openjdk.skara.bot.*;
+import org.openjdk.skara.host.*;
+import org.openjdk.skara.vcs.*;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Files;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Bot that automatically merges any changes from a dependency branch into a target branch
+ */
+class TopologicalBot implements Bot, WorkItem {
+    private final Logger log = Logger.getLogger("org.openjdk.skara.bots");
+    private final Path storage;
+    private final HostedRepository hostedRepo;
+    private final List<Branch> branches;
+    private final String depsFileName;
+
+    TopologicalBot(Path storage, HostedRepository repo, List<Branch> branches, String depsFileName) {
+        this.storage = storage;
+        this.hostedRepo = repo;
+        this.branches = branches;
+        this.depsFileName = depsFileName;
+    }
+
+    @Override
+    public boolean concurrentWith(WorkItem other) {
+        if (!(other instanceof TopologicalBot)) {
+            return true;
+        }
+        var otherBot = (TopologicalBot) other;
+        return !hostedRepo.getName().equals(otherBot.hostedRepo.getName());
+    }
+
+    @Override
+    public void run(Path scratchPath) {
+        log.info("Starting topobot run");
+        try {
+            var sanitizedUrl = URLEncoder.encode(hostedRepo.getWebUrl().toString(), StandardCharsets.UTF_8);
+            var dir = storage.resolve(sanitizedUrl);
+            Repository repo;
+            if (!Files.exists(dir)) {
+                log.info("Cloning " + hostedRepo.getName());
+                Files.createDirectories(dir);
+                repo = Repository.clone(hostedRepo.getUrl(), dir);
+            } else {
+                log.info("Found existing scratch directory for " + hostedRepo.getName());
+                repo = Repository.get(dir)
+                        .orElseThrow(() -> new RuntimeException("Repository in " + dir + " has vanished"));
+            }
+
+            repo.fetchAll();
+            var depsFile = repo.root().resolve(depsFileName);
+
+            List<Branch> ordered = orderedBranches(repo, depsFile);
+            log.info("Merge order " + ordered);
+            for (Branch branch : ordered) {
+                log.info("Processing branch " + branch + "...");
+                repo.checkout(branch);
+                Set<String> parents = new HashSet<>(
+                        Files.exists(depsFile) ? Files.readAllLines(depsFile) : List.of("master"));
+                List<String> failedMerges = new ArrayList<>();
```

Comment:
Isn't the format of the `deps.txt` file that the branches are separated with space? So maybe this should be:
```java
Set<String> parents = null;
if (Files.exists(depsFIle)) {
    var lines = Files.readAllLines(depsFile).stream().filter(String::isEmpty).collect(Collectors.toList());
    if (lines.size() > 1) {
        throw new IllegalStateException("Multiple non-empty lines in " + depsFile.toString() + ": " + String.join("\n", lines));
    }
    parents = new HashSet<>(Arrays.asList(lines.get(0).split(" ")));
} else {
    parents = Set.of("master");
}

Reply:
You're right, thanks for catching that.